api/websocket/messagebuffer/messagebuffer.go

```diff
@@ -2,38 +2,236 @@ package messagebuffer
+	// add it to index tree
+	idxNode := &indexNode{clientID: clientIDStr, expireId: expireId, createId: createId}
+	messageBuffer.treeCreate.Put(createId, idxNode)
+	messageBuffer.treeExpire.Put(expireId, idxNode)
+
 }
 
 // PopMessages reads and clears all messages of a client
 func (messageBuffer *MessageBuffer) PopMessages(clientID []byte) []*pb.Relay {
 	clientIDStr := hex.EncodeToString(clientID)
+
 	messageBuffer.Lock()
 	defer messageBuffer.Unlock()
-	messages := messageBuffer.buffer[clientIDStr]
-	messageBuffer.buffer[clientIDStr] = nil
-	return messages
+
+	treeMsg := messageBuffer.buffer[clientIDStr]
+	n := treeMsg.Size()
+
+	relayMessage := make([]*pb.Relay, n, n)
+	i := 0
+	for node := treeMsg.Left(); node != nil; node = treeMsg.Left() { // delete them from index tree
+		mNode := node.Value.(*msgNode)
+		treeMsg.Remove(mNode.createId)
+		messageBuffer.treeCreate.Remove(mNode.createId)
+		messageBuffer.treeExpire.Remove(mNode.expireId)
+		messageBuffer.totalBufSize -= mNode.GetSize()
+
+		relayMessage[i] = mNode.msg
+		i++
+	}
+
+	return relayMessage
+}
+
+// remove expired message
+func (messageBuffer *MessageBuffer) removeExpired() (removedCount int) {
+	now := time.Now().UnixMilli() + 1 // plus 1 milli-second
+	nowId := now * maxSeqId
+
+	removedCount = 0
+	minNode := messageBuffer.treeExpire.Left()
+	for minNode != nil {
+		idxNode := minNode.Value.(*indexNode)
+
+		if idxNode.expireId < nowId { // expired
+			v, ok := messageBuffer.buffer[idxNode.clientID].Get(idxNode.createId)
+			if ok { // remove from buffer
+				mNode := v.(*msgNode)
+				messageBuffer.buffer[idxNode.clientID].Remove(idxNode.createId)
+				messageBuffer.totalBufSize -= mNode.GetSize()
+			}
+			// remove from index tree
+			messageBuffer.treeExpire.Remove(idxNode.expireId)
+			messageBuffer.treeCreate.Remove(idxNode.createId)
+			removedCount++
+		} else {
+			break
+		}
+		minNode = messageBuffer.treeExpire.Left()
+	}
+
+	// printTree(os.Stdout, treeExpire.Root, 0, 'M')
+	return
+}
+
+// remove oldest message
+func (messageBuffer *MessageBuffer) removeOldest() {
+	minNode := messageBuffer.treeCreate.Left()
+	if minNode == nil { // the tree is empty
+		return
+	}
+
+	createId := minNode.Key.(int64)
+	idxNode := minNode.Value.(*indexNode)
+	treeMsg, ok := messageBuffer.buffer[idxNode.clientID]
+	if ok {
+		v, ok := treeMsg.Get(createId)
+		if ok {
+			treeMsg.Remove(createId)
+			mNode := v.(*msgNode)
+			messageBuffer.totalBufSize -= mNode.GetSize()
+		}
+
+		messageBuffer.treeExpire.Remove(idxNode.expireId)
+		messageBuffer.treeCreate.Remove(createId)
+	}
+
+}
+
+// monitor message buffer usage, and remove expired message .
+func (messageBuffer *MessageBuffer) BufferMonitor() {
+	for {
+		time.Sleep(10 * time.Second)
+		// delete expired msg
+		messageBuffer.removeExpired()
+		// still over size
+		for messageBuffer.totalBufSize > int(config.Parameters.ClientMsgCacheSize) { // repeat until totalBufSize less than MaxClientMsgCacheSize
```

Comment:
This line also need to be protected by lock. You can lock before `removeExpired()` and unlock after this for loop

Reply:
to avoid lock in lock. I use lock here seperately from removeExpired and removeOldest.